File: 01. Creacionales/03. Abstract Factory/01. abstractFactory.js
/***

Imaginemos que estamos construyendo una aplicación de compras en línea y necesitamos manejar diferentes 
tipos de productos de supermercado, como alimentos, productos de limpieza y artículos de cuidado personal. 
Cada tipo de producto tiene diferentes variantes, como diferentes marcas o tamaños.

Aquí es donde el patrón Abstract Factory puede ser útil. 
Podemos utilizar el patrón Abstract Factory para proporcionar una interfaz común para crear diferentes familias
 de productos relacionados, como alimentos, productos de limpieza y artículos de cuidado personal, 
 y luego implementar fábricas concretas para cada familia de productos.

*/

// Comencemos definiendo nuestra interfaz común para los productos de supermercado:

class Product {
    constructor(name, price) {
        this.name = name;
        this.price = price;
    }

    getInfo() {
        throw new Error("Método getInfo() debe ser implementado por las subclases.");
    }
}

// A continuación, definimos nuestras clases concretas que implementan la interfaz Product para cada tipo de producto:

class FoodProduct extends Product {
    constructor(name, price, brand) {
        super(name, price);
        this.brand = brand;
    }

    getInfo() {
        return `Alimento: ${this.name} - Marca: ${this.brand} - Precio: ${this.price}`;
    }
}

class CleaningProduct extends Product {
    constructor(name, price, size) {
        super(name, price);
        this.size = size;
    }

    getInfo() {
        return `Producto de limpieza: ${this.name} - Tamaño: ${this.size} - Precio: ${this.price}`;
    }
}

class PersonalCareProduct extends Product {
    constructor(name, price, type) {
        super(name, price);
        this.type = type;
    }

    getInfo() {
        return `Artículo de cuidado personal: ${this.name} - Tipo: ${this.type} - Precio: ${this.price}`;
    }
}

// Luego, definimos nuestra fábrica abstracta ProductFactory que proporciona métodos para crear diferentes tipos de productos:

class ProductFactory {
    createFoodProduct(name, price, brand) {
        throw new Error("Método createFoodProduct() debe ser implementado por las subclases.");
    }

    createCleaningProduct(name, price, size) {
        throw new Error("Método createCleaningProduct() debe ser implementado por las subclases.");
    }

    createPersonalCareProduct(name, price, type) {
        throw new Error("Método createPersonalCareProduct() debe ser implementado por las subclases.");
    }
}

// Finalmente, implementamos fábricas concretas que extienden la fábrica abstracta ProductFactory para cada familia de productos:

class SupermarketAFactory extends ProductFactory {
    createFoodProduct(name, price, brand) {
        return new FoodProduct(name, price, brand);
    }

    createCleaningProduct(name, price, size) {
        return new CleaningProduct(name, price, size);
    }

    createPersonalCareProduct(name, price, type) {
        return new PersonalCareProduct(name, price, type);
    }
}

class SupermarketBFactory extends ProductFactory {
    createFoodProduct(name, price, brand) {
        return new FoodProduct(name, price, brand);
    }

    createCleaningProduct(name, price, size) {
        return new CleaningProduct(name, price, size);
    }

    createPersonalCareProduct(name, price, type) {
        return new PersonalCareProduct(name, price, type);
    }
}

// Ahora podemos utilizar el patrón Abstract Factory para crear diferentes productos de supermercado utilizando la fábrica correspondiente:

const supermarketAFactory = new SupermarketAFactory();
const foodProductA = supermarketAFactory.createFoodProduct("Arroz", 2.99, "Marca A");
console.log(foodProductA.getInfo()); // Alimento: Arroz - Marca: Marca A - Precio: 2.99

const supermarketBFactory = new SupermarketBFactory();
const cleaningProductB = supermarketBFactory.createCleaningProduct("Detergente", 1.99, "500ml");
console.log(cleaningProductB.getInfo()); // Producto de limpieza: Detergente - Tamaño: 500ml - Precio: 1.99

const personalCareProductB = supermarketBFactory.createPersonalCareProduct("Jabón", 0.99, "Barra");
console.log(personalCareProductB.getInfo());

/***

Ahora hemos utilizado el patrón Abstract Factory para crear diferentes productos de supermercado utilizando 
las fábricas correspondientes. Esto nos permite crear fácilmente productos de diferentes familias 
sin preocuparnos por los detalles de implementación de cada uno.

*/

File: 01. Creacionales/03. Abstract Factory/01. abstractFactory.py
from abc import ABC, abstractmethod

# Comencemos definiendo nuestra interfaz común para los productos de supermercado:
class Product(ABC):
    def __init__(self, name, price):
        self.name = name
        self.price = price

    @abstractmethod
    def get_info(self):
        pass

# A continuación, definimos nuestras clases concretas que implementan la interfaz Product para cada tipo de producto:
class FoodProduct(Product):
    def __init__(self, name, price, brand):
        super().__init__(name, price)
        self.brand = brand

    def get_info(self):
        return f"Alimento: {self.name} - Marca: {self.brand} - Precio: {self.price}"

class CleaningProduct(Product):
    def __init__(self, name, price, size):
        super().__init__(name, price)
        self.size = size

    def get_info(self):
        return f"Producto de limpieza: {self.name} - Tamaño: {self.size} - Precio: {self.price}"

class PersonalCareProduct(Product):
    def __init__(self, name, price, type):
        super().__init__(name, price)
        self.type = type

    def get_info(self):
        return f"Artículo de cuidado personal: {self.name} - Tipo: {self.type} - Precio: {self.price}"

# Luego, definimos nuestra fábrica abstracta ProductFactory que proporciona métodos para crear diferentes tipos de productos:
class ProductFactory(ABC):
    @abstractmethod
    def create_food_product(self, name, price, brand):
        pass

    @abstractmethod
    def create_cleaning_product(self, name, price, size):
        pass

    @abstractmethod
    def create_personal_care_product(self, name, price, type):
        pass

# Finalmente, implementamos fábricas concretas que extienden la fábrica abstracta ProductFactory para cada familia de productos:
class SupermarketAFactory(ProductFactory):
    def create_food_product(self, name, price, brand):
        return FoodProduct(name, price, brand)

    def create_cleaning_product(self, name, price, size):
        return CleaningProduct(name, price, size)

    def create_personal_care_product(self, name, price, type):
        return PersonalCareProduct(name, price, type)

class SupermarketBFactory(ProductFactory):
    def create_food_product(self, name, price, brand):
        return FoodProduct(name, price, brand)

    def create_cleaning_product(self, name, price, size):
        return CleaningProduct(name, price, size)

    def create_personal_care_product(self, name, price, type):
        return PersonalCareProduct(name, price, type)

# Ahora podemos utilizar el patrón Abstract Factory para crear diferentes productos de supermercado utilizando la fábrica correspondiente:
supermarket_a_factory = SupermarketAFactory()
food_product_a = supermarket_a_factory.create_food_product("Arroz", 2.99, "Marca A")
print(food_product_a.get_info()) # Alimento: Arroz - Marca: Marca A - Precio: 2.99

supermarket_b_factory = SupermarketBFactory()
cleaning_product_b = supermarket_b_factory.create_cleaning_product("Detergente", 1.99, "500ml")
print(cleaning_product_b.get_info()) # Producto de limpieza: Detergente - Tamaño: 500ml - Precio: 1.99

personal_care_product_b = supermarket_b_factory.create_personal_care_product("Jabón", 0.99, "Barra")
print(personal_care_product_b.get_info())